### PR TITLE
Return the first consecutive sub block of ops for frs driver instead of 0 ops on violation

### PR DESCRIPTION
**Pull request by jatgarg**
## Description

Item: https://dev.azure.com/fluidframework/internal/_workitems/edit/4668
When we fetch ops for frs through delta storage or through snapshot, it could happen that the ops are not contiguous. In that case, return the first set of contiguous ops so that we can make at least some progress.

It's an issue service needs to address, but this change is a temp experiment to see if it mitigates issues for users while we are waiting for proper fix.

### CODE REVIEW

**Comment by vladsud** [comment]
I'd add to your description that it's an issue service needs to address, but this change is a temp experiment to see if it mitigates issues for users while we are waiting for proper fix.

File: packages/drivers/routerlicious-driver/src/deltaStorageService.ts [hunk 1]
```diff
@@ -73,7 +73,7 @@ export class DocumentDeltaStorageService implements IDocumentDeltaStorageService
 				const messages = this.snapshotOps.filter(
 					(op) => op.sequenceNumber >= from && op.sequenceNumber < to,
 				);
-				validateMessages("snapshotOps", messages, from, this.logger);
+				validateMessages("snapshotOps", messages, from, this.logger, false);
```
Comment (vladsud) [inline packages/drivers/routerlicious-driver/src/deltaStorageService.ts:76]: , false /* strict /*)

File: packages/drivers/driver-base/src/driverUtils.ts [hunk 2]
```diff
@@ -128,16 +129,38 @@ export function validateMessages(
...
+			if (strict) {
+				messages.length = 0;
+			} else {
+				let validOpsCount = 1;
+				while (
+					validOpsCount < messages.length &&
+					messages[validOpsCount].sequenceNumber ===
+						messages[validOpsCount - 1].sequenceNumber + 1
+				) {
+					validOpsCount += 1;
```
Comment (vladsud) [inline packages/drivers/driver-base/src/driverUtils.ts:142]: validOpsCount ++

File: packages/drivers/driver-base/src/driverUtils.ts [hunk 3]
```diff
@@ -128,16 +129,38 @@ export function validateMessages(
...
+				}
+				messages.length = Math.min(messages.length, validOpsCount);
```
Comment (vladsud) [inline packages/drivers/driver-base/src/driverUtils.ts:144]: You do not need to do min() here, validOpsCount is guaranteed to be not higher.

File: packages/drivers/driver-base/src/driverUtils.ts [hunk 4]
```diff
@@ -128,16 +129,38 @@ export function validateMessages(
...
+			}
+			let firstValidOpSeqNumber: number | undefined;
+			let lastValidOpSeqNumber: number | undefined;
+			if (messages.length > 0) {
+				firstValidOpSeqNumber = messages[0].sequenceNumber;
```
Comment (vladsud) [inline packages/drivers/driver-base/src/driverUtils.ts:149]: I'm not sure logging firstValidOpSeqNumber is useful - it's always same as `from`.
I'd instead have too lengths - original one, and one after fixup

**Comment by vladsud** [comment]
Q: any other places to use validateMessages()? I believe we were hitting (in the past) asserts somewhere in GetOps flow (in shared code), and code there was doing pretty much same validation (but using asserts).

Maybe worth glancing at code around 0x26d, 0x26e, 0x26f. I'm not sure what changed and why we are not hitting them anymore (maybe FRS fixed issues here? Current set of problems seems to be only with ops that are part of snapshot).
If so, maybe it's fine to leave it as is (at least for now).

File: packages/drivers/driver-base/src/driverUtils.ts (hunk 3 above)
Comment (jatgarg) [reply packages/drivers/driver-base/src/driverUtils.ts:144]: In corner case when the messages only had 0/1 ops and validOpsCount start with 1. So I did the min here.

**Comment by jatgarg** [comment]
> Q: any other places to use validateMessages()? I believe we were hitting (in the past) asserts somewhere in GetOps flow (in shared code), and code there was doing pretty much same validation (but using asserts).
> 
> Maybe worth glancing at code around 0x26d, 0x26e, 0x26f. I'm not sure what changed and why we are not hitting them anymore (maybe FRS fixed issues here? Current set of problems seems to be only with ops that are part of snapshot). If so, maybe it's fine to leave it as is (at least for now).

As mentioned in the issue item, here are the findings. We seem to hit "OpViolation" in both snapshot ops and in storage ops.
<div style="box-sizing: border-box; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial; color: rgb(0, 0, 0); background-color: rgb(255, 255, 254); font-family: Menlo, Monaco, &quot;Courier New&quot;, monospace; font-size: 12px; line-height: 18px;"><div style="box-sizing: border-box;"><span style="box-sizing: border-box;">let opsViolation = union office_fluid_ffautomation*</span></div><div style="box-sizing: border-box;"><span style="box-sizing: border-box;">| where Event_Time &gt; ago(30d)</span></div><div style="box-sizing: border-box;"><span style="box-sizing: border-box;">| where Data_hostName == "@fluid-internal/test-service-load"</span></div><div style="box-sizing: border-box;"><span style="box-sizing: border-box;">| where Data_branch startswith "refs/heads/main"</span></div><div style="box-sizing: border-box;"><span style="box-sizing: border-box;">| where Data_driverEndpointName contains "frs"</span></div><div style="box-sizing: border-box;"><span style="box-sizing: border-box;">| where Data_eventName has "OpsFetchViolation"</span></div><div style="box-sizing: border-box;"><span style="box-sizing: border-box;">| summarize count() by Data_reason;</span></div><div style="box-sizing: border-box;"><span style="box-sizing: border-box;">//| project Data_eventName, Data_reason, Data_from, Data_start, Data_last, Data_length;</span></div><div style="box-sizing: border-box;"><span style="box-sizing: border-box;">//| summarize count() by bin(Event_Time, 1d);</span></div><div style="box-sizing: border-box;"><span style="box-sizing: border-box;">opsViolation</span></div></div><div style="box-sizing: border-box; color: rgba(0, 0, 0, 0.9); font-family: &quot;Segoe UI VSS (Regular)&quot;, &quot;Segoe UI&quot;, -apple-system, &quot;system-ui&quot;, Roboto, &quot;Helvetica Neue&quot;, Helvetica, Ubuntu, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;, &quot;Segoe UI Symbol&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><br style="box-sizing: border-box;"><br style="box-sizing: border-box;">Table0

Data_reason | count_
-- | --
snapshotOps | 15,547
storage | 9,984

</div><br class="Apple-interchange-newline">

I am following up on this with @MaggieLiMS and they will be fixing this at their end.

**Comment by jatgarg** [comment]
However, these asserts are not present anymore for last 30 days. However, we will catch them here and not cause those asserts any more it seems. Also we have this item opened here: https://dev.azure.com/fluidframework/internal/_workitems/edit/4667
I will be looking into those asserts as part of above item.

File: packages/drivers/driver-base/src/driverUtils.ts (hunk 3 above)
Comment (vladsud) [reply packages/drivers/driver-base/src/driverUtils.ts:134]: If length == 0, then we would not get here (check on line 115). In all other cases validOpsCount would be less or equal to messages.length (condition on like 127). So I think you can safely remove min()

**Review by vladsud — APPROVED** [review]

File: packages/drivers/driver-base/src/driverUtils.ts (hunk 3 above)
Comment (jatgarg) [reply packages/drivers/driver-base/src/driverUtils.ts:134]: Created PR here: https://github.com/microsoft/FluidFramework/pull/16206
It was required previously when we had separate ifs :)

**Pull request merged**